base_layer/mmr/benches/bench.rs

```diff
@@ -22,29 +22,25 @@
 //
 
 use blake2::Blake2b;
-use criterion::{criterion_group, criterion_main, Criterion};
+use criterion::{criterion_group, criterion_main, BatchSize, Criterion};
 use digest::Digest;
 use std::time::Duration;
 use tari_mmr::MerkleMountainRange;
 
-fn get_hashes(n: usize) -> Vec<Vec<u8>> {
-    let mut result = Vec::with_capacity(n);
-    for i in 0..n {
-        let h = Blake2b::digest(&i.to_le_bytes()).to_vec();
-        result.push(h);
-    }
-    result
-}
-
 fn build_mmr(c: &mut Criterion) {
     c.bench_function("Build MMR", move |b| {
-        let hashes = get_hashes(1000);
+        let mut counter = 0usize;
         let mut mmr = MerkleMountainRange::<Blake2b, _>::new(Vec::default());
-        b.iter(|| {
-            for i in 0..1000 {
-                let _ = mmr.push(&hashes[i]);
-            }
-        });
+        b.iter_batched(
+            || {
+                counter += 1;
+                Blake2b::digest(&counter.to_le_bytes()).to_vec()
+            },
+            |hash| {
+                mmr.push(hash).unwrap();
+            },
+            BatchSize::SmallInput,
+        );
```

Comment:
You changed the benchmark here from one that measures the aggregate performance of many inserts to one that measures a single insert.

But not all inserts are equal. Some just add the hash, while others have to a bunch of peak calculations. So this means that the benchmark cannot create an apples for apples set of results and will always return a lot of outliers.

A more accurate bench would be use use the same idea as before (a 1000 hash MMR), but still use `iter_batch`

```
fn get_hashes(n: usize) -> Vec<Vec<u8>> {
    let mut result = Vec::with_capacity(n);
    for i in 0..n {
        let h = Blake2b::digest(&i.to_le_bytes()).to_vec();
        result.push(h);
    }
    result
}

fn build_mmr(c: &mut Criterion) {
    c.bench_function("Build MMR", move |b| {
        let hashes = get_hashes(1000);
        let mut mmr = MerkleMountainRange::<Blake2b, _>::new(Vec::default());
        b.iter_batched(
            || hashes.clone(),
            |hashes| {
                hashes.into_iter().for_each(|hash| { mmr.push(hash).unwrap(); });
            },
            BatchSize::SmallInput,
        );
    });
}
```

Reply:
Yes that makes sense - code updated